packages/grid/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx

```diff
@@ -0,0 +1,220 @@
+
+const CURSOR_MOVE_DIRECTION_TOP = 'top';
+const CURSOR_MOVE_DIRECTION_BOTTOM = 'bottom';
+
+const getCursorMoveDirectionY = (
+  currentCoordinates: CursorCoordinates,
+  nextCoordinates: CursorCoordinates,
+) => {
+  return currentCoordinates.y <= nextCoordinates.y
+    ? CURSOR_MOVE_DIRECTION_BOTTOM
+    : CURSOR_MOVE_DIRECTION_TOP;
+};
+
+const hasCursorPositionChanged = (
+  currentCoordinates: CursorCoordinates,
+  nextCoordinates: CursorCoordinates,
+): boolean =>
+  currentCoordinates.x !== nextCoordinates.x || currentCoordinates.y !== nextCoordinates.y;
+
+type OwnerState = { classes: DataGridProProcessedProps['classes'] };
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { classes } = ownerState;
+
+  const slots = {
+    rowDragging: ['row--dragging'],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+export const rowReorderStateInitializer: GridStateInitializer = (state) => ({
+  ...state,
+  rowReorder: { dragRow: '' },
+});
+
+/**
+ * Only available in DataGridPro
+ * @requires useGridRows (method)
+ */
+export const useGridRowReorder = (
+  apiRef: React.MutableRefObject<GridApiPro>,
+  props: Pick<DataGridProProcessedProps, 'disableRowReorder' | 'onRowOrderChange' | 'classes'>,
+): void => {
+  const logger = useGridLogger(apiRef, 'useGridRowReorder');
+  const sortModel = useGridSelector(apiRef, gridSortModelSelector);
+  const treeDepth = useGridSelector(apiRef, gridRowTreeDepthSelector);
+  const dragRowNode = React.useRef<HTMLElement | null>(null);
+  const originRowIndex = React.useRef<number | null>(null);
+  const removeDnDStylesTimeout = React.useRef<any>();
+  const cursorPosition = React.useRef<CursorCoordinates>({
+    x: 0,
+    y: 0,
+  });
+  const ownerState = { classes: props.classes };
+  const classes = useUtilityClasses(ownerState);
+
+  React.useEffect(() => {
+    return () => {
+      clearTimeout(removeDnDStylesTimeout.current);
+    };
+  }, []);
+
+  // TODO: remove sortModel check once row reorder is sorting compatible
+  // remove treeDepth once row reorder is tree compatible
+  const isRowReorderDisabled = React.useCallback((): boolean => {
+    return props.disableRowReorder || !!sortModel.length || treeDepth !== 1;
+  }, [props.disableRowReorder, sortModel, treeDepth]);
+
+  const handleDragStart = React.useCallback<GridEventListener<GridEvents.rowDragStart>>(
+    (params, event) => {
+      if (isRowReorderDisabled()) {
+        return;
+      }
+
+      logger.debug(`Start dragging row ${params.id}`);
+      // Prevent drag events propagation.
+      // For more information check here https://github.com/mui/mui-x/issues/2680.
+      event.stopPropagation();
+
+      dragRowNode.current = event.currentTarget;
+      dragRowNode.current.classList.add(classes.rowDragging);
+
+      apiRef.current.setState((state) => ({
+        ...state,
+        rowReorder: { ...state.rowReorder, dragRow: params.id },
```

Comment:
In https://github.com/mui/mui-x/issues/820#issuecomment-897906608 I proposed to clean a bit the state and replace some keys with internal state. This would include `columnReorder`. Could we use a simple `React.useState` here? We only need to track which row is being dragged.